In [2]:
from google.colab import drive
drive.mount('/content/drive')

ModuleNotFoundError: No module named 'google.colab'

In [1]:
# Importar las librerías con el código desarrollado
import sys
sys.path.insert(0,'/content/drive/MyDrive/Colab Notebooks/MLOPS')

In [ ]:
# Librerías a usar en el proyecto
from download_data import get_data
import pandas as pd
import plotly.express as px

Mounted at /content/drive
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


In [ ]:
# Obtención de los archivos
news_true, news_false = get_data()

# **1. Análisis del archivo noticias reales**
De acuerdo con el análisis exploratorio de los datos se encuentra:
1. Total de noticias: 21417
2. Noticias en el año 2016: 4716
3. Noticias en el año 2017: 16701
4. Las noticias se encuentran asociadas a dos temas: 'politicsNews' y 'worldnews'. 
5. Noticias que tratan de politicsNews : 11272
6. Noticias que tratan de worldnews: 10145

In [ ]:
# Conteo del número de registros
news_true.count()

title      21417
text       21417
subject    21417
date       21417
dtype: int64

In [ ]:
# Información del dataframe
news_true.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB


In [ ]:
# Convertir el formato de la columna date a datetime
news_true['date'] = pd.to_datetime(news_true['date'], infer_datetime_format=True)

In [ ]:
# Extraer el año de la fecha
news_true['anno'] = news_true['date'].dt.year

In [ ]:
# Validación de la conversión del formato y creación de los nuevos elementos 
news_true.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   title    21417 non-null  object        
 1   text     21417 non-null  object        
 2   subject  21417 non-null  object        
 3   date     21417 non-null  datetime64[ns]
 4   anno     21417 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 836.7+ KB


In [ ]:
# Número de noticias por año
dfg=news_true.groupby('anno').count().reset_index()
fig = px.bar(dfg, x='anno', y = 'title', title = 'Número de noticias reales por año', color = 'title', barmode = 'stack', text_auto='.2s',
              labels={'anno':'Año de la noticia','title':'Número de noticias'})
fig.show()

In [ ]:
# Análisis del subject de las noticias
print(news_true['subject'].unique())

['politicsNews' 'worldnews']


In [ ]:
# Número de noticias por tópico
dfg=news_true.groupby('subject').count().reset_index()
fig = px.bar(dfg, x='subject', y = 'title', title = 'Número de noticias reales por tópico', color = 'title', barmode = 'stack',
              labels={'subject':'Tópico de la noticia','title':'Número de noticias'})
fig.show()

In [ ]:
# Análisis del contenido de las noticias tanto en el título como en el texto
news_true.iloc[15001, 1]

'PANAMA CITY (Reuters) - Panama s President Juan Carlos Varela will travel to China to meet with his counterpart Xi Jinping on Friday, the first official visit by a Panamanian leader to the Asian country  coming five months after the nations established diplomatic relations. The bilateral meeting will serve to  establish a new economic, trade, tourist and diplomatic outlook for the country, leading to more than a dozen agreements that will be signed  between nations, Panama s Foreign Ministry said in a statement. In June, Panama upgraded its commercial ties with China and established full diplomatic links with the second most important customer of its key shipping canal, giving Beijing a major victory as it broke formal relations with Taiwan. The Foreign Ministry said the agreements will provide the basis for attracting investments and Chinese innovation to Panama, and help boost Panamanian exports, tourism and the use of the Panama canal. Varela will inaugurate Panama s first embassy 

# **2. Análisis del archivo noticias falsas**
De acuerdo con el análisis exploratorio de los datos se encuentra:
1. Total de noticias: 23481.
2. El campo 'date' presenta diferentes formatos e incluso datos que no corresponden a la fecha, sin embargo dado el problema que se plantea y como resultado de análisis exploratorio no se considera información relevante para el desarrollo del objetivo del proyecto.
4. Las noticias se encuentran asociadas a dos temas: 'News', 'politics', 'Government News', 'left-news', 'US_News' y 'Middle-east.
5. Noticias que tratan de politics : 6841.
6. Noticias que tratan de News: 9050.
7. Noticias que tratan de Government News: 1570.
8. Noticias que tratan de left-News: 4459.
9. Noticias que tratan de US_News: 783.
10. Noticias que tratan de Middle-east: 778.

In [ ]:
# Conteo del número de registros
news_false.count()

title      23481
text       23481
subject    23481
date       23481
dtype: int64

In [ ]:
# Información del dataframe
news_false.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


In [ ]:
# Análisis del subject de las noticias
print(news_false['subject'].unique())

['News' 'politics' 'Government News' 'left-news' 'US_News' 'Middle-east']


In [ ]:
# Número de noticias por tópico
dfg=news_false.groupby('subject').count().reset_index()
fig = px.bar(dfg, x='subject', y = 'title', title = 'Número de noticias reales por tópico', color = 'title', barmode = 'stack',
              labels={'subject':'Tópico de la noticia','title':'Número de noticias'})
fig.show()

In [ ]:
# Análisis exploratorio del campo date, al encontrarse que no todas las filas tienen una fecha en esa columna
serie = news_false.iloc[9350:9400,3]
print(serie)

9350                                         Nov 26, 2017
9351                                         Nov 26, 2017
9352                                         Nov 26, 2017
9353                                         Nov 25, 2017
9354                                         Nov 25, 2017
9355                                         Nov 25, 2017
9356                                         Nov 25, 2017
9357                                         Nov 25, 2017
9358    https://100percentfedup.com/served-roy-moore-v...
9359                                         Nov 25, 2017
9360                                         Nov 24, 2017
9361                                         Nov 24, 2017
9362                                         Nov 24, 2017
9363                                         Nov 23, 2017
9364                                         Nov 23, 2017
9365                                         Nov 22, 2017
9366                                         Nov 22, 2017
9367          

In [ ]:
# Ejemplo de uno de los registros y lo que aparece registrado en ese campo
serie9358 = news_false.iloc[9358,:]
print(serie9358[3])

https://100percentfedup.com/served-roy-moore-vietnamletter-veteran-sets-record-straight-honorable-decent-respectable-patriotic-commander-soldier/


In [ ]:
# Convertir el formato de la columna date a datetime
news_false['date_new'] = pd.to_datetime(news_false['date'], infer_datetime_format=True, errors='coerce')

In [ ]:
news_false.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   title     23481 non-null  object        
 1   text      23481 non-null  object        
 2   subject   23481 non-null  object        
 3   date      23481 non-null  object        
 4   date_new  11868 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(4)
memory usage: 917.4+ KB


In [ ]:
news_false['date_new'].unique()

array(['2017-12-31T00:00:00.000000000', '2017-12-30T00:00:00.000000000',
       '2017-12-29T00:00:00.000000000', '2017-12-25T00:00:00.000000000',
       '2017-12-23T00:00:00.000000000', '2017-12-22T00:00:00.000000000',
       '2017-12-21T00:00:00.000000000', '2017-12-20T00:00:00.000000000',
       '2017-12-19T00:00:00.000000000', '2017-12-17T00:00:00.000000000',
       '2017-12-16T00:00:00.000000000', '2017-12-13T00:00:00.000000000',
       '2017-12-12T00:00:00.000000000', '2017-12-11T00:00:00.000000000',
       '2017-12-10T00:00:00.000000000', '2017-12-09T00:00:00.000000000',
       '2017-12-07T00:00:00.000000000', '2017-12-06T00:00:00.000000000',
       '2017-12-05T00:00:00.000000000', '2017-12-04T00:00:00.000000000',
       '2017-12-03T00:00:00.000000000', '2017-12-02T00:00:00.000000000',
       '2017-12-01T00:00:00.000000000', '2017-11-30T00:00:00.000000000',
       '2017-11-29T00:00:00.000000000', '2017-11-28T00:00:00.000000000',
       '2017-11-27T00:00:00.000000000', '2017-11-26

In [ ]:
result = news_false[news_false.date_new.isnull()]
result.head(0)

title  \
9050  Democrat Senator Warns Mueller Not To Release ...   
9051  MSNBC ANCHOR Flabbergasted at What Texas Teach...   
9052  WATCH: SNOWFLAKES ASKED Communist Party Platfo...   
9053  JUST IN: BADASS GENERAL JOHN KELLY Shoved Chin...   
9054  DOJ’s JEFF SESSIONS Opens Investigation Into W...   
...                                                 ...   
9345  NORWAY’S TOUGH Immigration Minister BLASTS Swe...   
9346  BRITISH MUSICIAN Says If He Was Given The Chan...   
9347  LOL! DEMOCRAT CHEERLEADER Cokie Roberts Delive...   
9348  SOUTH KOREA Uses Hilarious Technique To TAUNT ...   
9349  88-Yr Old DEMOCRAT Congressman and Accused SEX...   

                                                   text   subject  \
9050  According to The Hill, Democrat Senator Bob Ca...  politics   
9051  If we protect every other government building ...  politics   
9052  Ami Horowitz is fantastic! Check out this man ...  politics   
9053  Just one more reminder of why President Trump ...  politics   
9054  Thank goodnesss Jeff Sessions is moving on fin...  politics   
...                                                 ...       ...   
9345  Norway has a new immigration sheriff in town, ...  politics   
9346  If this so-called celebrity said that he would...  politics   
9347  Earlier today, President Trump tweeted about t...  politics   
9348  There is no word yet, about how Rocket Man pla...  politics   
9349  John Conyers, the 88-year old Democrat Congres...  politics   

              date date_new  
9050     19-Feb-18      NaT  
9051     19-Feb-18      NaT  
9052     19-Feb-18      NaT  
9053     18-Feb-18      NaT  
9054     18-Feb-18      NaT  
...            ...      ...  
9345  Nov 27, 2017      NaT  
9346  Nov 26, 2017      NaT  
9347  Nov 26, 2017      NaT  
9348  Nov 26, 2017      NaT  
9349  Nov 26, 2017      NaT  

[300 rows x 5 columns]

In [ ]:
serietest = news_false.iloc[9049,:]
print(serietest)

title        Ben Carson Campaign In Shambles After Top Aid...
text        It s not looking like much of a Happy New for ...
subject                                                  News
date                                          January 1, 2016
date_new                                  2016-01-01 00:00:00
Name: 9049, dtype: object
